File: Data Structure/Stack/Easy/LET_no20_Valid Parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        if len(s) % 2 != 0:
            return False
        
        memo = {'[':']', '{':'}', '(':')'}
        for c in s:
            if c in "[({":
                stack.append(c)
            else:
                if not stack:
                    return False
                if memo[stack[-1]] != c:
                    return False
                else:
                    stack.pop()
        
        return len(stack) == 0

File: Data Structure/Stack/Easy/LET_no20_Valid Parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        vector<char> stk;
        for (char c : s) {
            if (!stk.empty()) {
                char c2 = stk.back();
                if (c2 == '(' && c == ')' ||
                    c2 == '{' && c == '}' ||
                    c2 == '[' && c == ']') 
                        stk.pop_back();
                else
                    stk.push_back(c);
            }
            else
                stk.push_back(c);
        }

        return stk.empty();
    }
};